### PR TITLE
added diffrentialable contacts option

### PR DESCRIPTION
**Pull request by msultan**
Added option to calculate contacts using a diffrentialable scheme when using the residue contact distance option.

### CODE REVIEW

**Comment by msultan** [comment]
Exact scheme is inspired from [here](http://plumed.github.io/doc-v2.0/user-doc/html/mindist.html)

File: mdtraj/geometry/contact.py [hunk 1]
```diff
@@ -40,7 +40,50 @@
 # Code
 ##############################################################################
 
-def compute_contacts(traj, contacts='all', scheme='closest-heavy', ignore_nonprotein=True, periodic=True):
+##############################################################################
+# MDTraj: A Python Library for Loading, Saving, and Manipulating
```
Comment (mpharrigan) [inline mdtraj/geometry/contact.py]: What happened here? Don't re-add the header

Comment (msultan) [inline mdtraj/geometry/contact.py]: yea, fixing it now

**Comment by mpharrigan** [comment]
This should be a new `scheme` not a new parameter. And it should probably be named "softmin" or something https://en.wikipedia.org/wiki/Softmax_function

**Comment by mpharrigan** [comment]
Or maybe it shouldn't be a new scheme because it can apply to each of the other schemes (except CA). Maybe add an exception if it's set to true and scheme is alpha carbon

Probably should be called `softmin` or something

File: mdtraj/geometry/contact.py [hunk 2]
```diff
@@ -71,6 +114,14 @@ def compute_contacts(traj, contacts='all', scheme='closest-heavy', ignore_nonpro
     periodic : bool, default=True
         If periodic is True and the trajectory contains unitcell information,
         we will compute distances under the minimum image convention.
+    diffrentialable_scheme : bool, default=False
+        If diffrentialable_scheme is true, we will use a diffrentiable version of
+        the scheme. The exact expression used
+         is d = \frac{\beta}{log\sum_i{exp(\frac{\beta}{d_i}})} where
+         beta is user parameter which defaults to 500
```
Comment (mpharrigan) [inline mdtraj/geometry/contact.py]: this is not the actual beta default

File: mdtraj/geometry/contact.py [hunk 3]
```diff
@@ -71,6 +114,14 @@ def compute_contacts(traj, contacts='all', scheme='closest-heavy', ignore_nonpro
...
+    beta : float, default=20
```
Comment (mpharrigan) [inline mdtraj/geometry/contact.py]: units? I think nm?

Comment (msultan) [inline mdtraj/geometry/contact.py]: yep, also good catch on the 500.

File: mdtraj/geometry/contact.py [hunk 4]
```diff
@@ -71,6 +114,14 @@ def compute_contacts(traj, contacts='all', scheme='closest-heavy', ignore_nonpro
     periodic : bool, default=True
         If periodic is True and the trajectory contains unitcell information,
         we will compute distances under the minimum image convention.
+    diffrentialable_scheme : bool, default=False
```
Comment (mpharrigan) [inline mdtraj/geometry/contact.py]: link to plumed user manual in the docs?

**Comment by msultan** [comment]
Thanks for all the feedback @mpharrigan

**Comment by msultan** [comment]
For CA, it doesnt matter if we use soft_min since the logsumexp cancels out. I dont think we need to explicitly encode that.

**Comment by mpharrigan** [comment]
> For CA, it doesnt matter if we use soft_min since the logsumexp cancels out. I dont think we need to explicitly encode that.

Maybe a warning? My thought is if people are doing those combinations of options, it's probably a mistake

File: mdtraj/geometry/contact.py [hunk 5]
```diff
@@ -218,7 +218,9 @@ def compute_contacts(traj, contacts='all', scheme='closest-heavy', ignore_nonpro
             if not soft_min:
                 distances[:, i] = atom_distances[:, index : index + n].min(axis=1)
             else:
-                distances[:, i] = beta / np.log(np.sum(np.exp(beta/atom_distances[:, index : index + n]), axis=1))
+                distances[:, i] = soft_min_beta / \
+                                  np.log(np.sum(np.exp(soft_min_beta/ \
```
Comment (mpharrigan) [inline mdtraj/geometry/contact.py]: this `\` is unnecessary

**Comment by mpharrigan** [comment]
errors seem unrelated

**Comment by msultan** [comment]
tests seem to be failing for some random reasons now.

**Pull request merged**